docs/decisions/0003-django-setting-toggles.rst

```diff
@@ -0,0 +1,66 @@
+Django Setting Toggles
+======================
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+Django Settings are implicitly defined when they are used.  For example, in ``edx-platform`` the ``FEATURES`` setting dict is used, and an example usage might look like::
+
+    settings.FEATURES.get('EXAMPLE_FEATURE', False)
+
+There are several problems with this style of implicit definition:
+
+* Each usage may or may not define a default value, and the default values are not guaranteed to be consistent.
+* It is unclear when and if a default value should be provided with the usage.
+* If there are multiple uses of a setting, it unclear where the code annotation that documents the toggle should live.
+* Django Settings toggles are unable to report the same level of detail in the toggle state report as Waffle toggles, because they can't be queried for location or code ownership information.
+
+Decision
+--------
+
+We will introduce a simple wrapping class that can be used to explicitly define a Django Setting toggle.
+
+Here is an example declaration::
+
+    EXAMPLE_FEATURE_TOGGLE = DjangoSettingToggle(
+        "FEATURES['EXAMPLE_FEATURE']",
+        default_value=False,
+    )
```

Comment:
There is one major issue with this implementation: currently, in edx-platform, there are several hundred settings (~700 by my last count if I remember correctly) and more than a hundred feature flags. Presumably, every setting is used at least once in the edx-platform codebase. We cannot imagine wrapping every single setting in a separate module.

I realize that this ADR is mostly about feature flags, and not meant all settings (or is it?); still, there are many feature flags in the current setting files.

Reply:
This is just about toggle settings at this point. I will mention that non-toggle settings are out-of-scope for this decision.
1. This doesn’t need to be rolled out immediately.
2. Another benefit is that we really want to clean up common.py, and this could be part of that clean-up, where appropriate.

Reply:
pouac